### PR TITLE
[mono] Add Vector128 Sum intrinsic for amd64

### PR DESCRIPTION
**Pull request by matouskozak**
Add support for the following Vector128 API's:

- Sum: It doesn't support byte and sbyte types yet. It does generate instructions for i64 type but not intrinsics but the assembly generated is significantly smaller than without it.

### CODE REVIEW

**Comment by jandupej** [comment]
I'm nitpicking here. For `f32`, this horizontal sum boils down to:
```
haddps xmm0, xmm0       ; ICL (p01 2p5) lat=6, thr=1/2  ; Zen3 lat=6 thr=1/2
haddps xmm0, xmm0       ; ICL (p01 2p5) lat=6, thr=1/2  ; Zen3 lat=6 thr=1/2
```

The `haddps` instruction has a latency of 6 both on ICL/TGL and Zen3. This could be slightly improved by eliminating the first `haddps`:

```
xorps xmm1, xmm1        ; ICL, Zen3 - dependency-breaker (probably lat=0)
movhlps xmm1, xmm0      ; ICL (p5) lat=1, thr=1         ; Zen3 lat=1, thr=2
addps xmm0, xmm1        ; ICL (p01) lat=4, thr=2        ; Zen3 lat=3, thr=2
haddps xmm0, xmm0       ; ICL (p01 2p5) lat=6, thr=1/2  ; Zen3 lat=6 thr=1/2
```

The resulting code is longer, but has a lower total latency and puts less pressure on Intel's port 5. 

Still, horizontal add probably won't be executed in an inner loop, so saving 1-2 clocks of latency is not significant. And this would probably have to be measured, too.

**Review by vargaz — APPROVED** [review]

**Comment by tannergooding** [comment]
> The resulting code is longer, but has a lower total latency and puts less pressure on Intel's port 5.

I expect the longer code will have an overall net-negative impact in loops since it takes up 2x the space, produces a 3 instruction dependency chain, and likewise will take up additional micro-ops in the decoder.

We also have to be careful because this can be non-deterministic if you aren't careful. For floating-point, `(a + b) + c` != `a + (b + c)` and so doing `a[0] + a[1] + a[2] + a[3]` for the scalar, but doing `(a[0] + a[1]) + (a[2] + a[3])` for 2x hadd or `(a[0] + [a2]) + (a[1] + a[3])` for shuffle, add, hadd; may all produce different results.

File: src/mono/mono/mini/simd-intrinsics.c [hunk 1]
```diff
@@ -545,6 +551,92 @@ emit_sum_vector (MonoCompile *cfg, MonoType *vector_type, MonoTypeEnum element_t
 }
 #endif
 
+#ifdef TARGET_AMD64
+static int type_to_extract_op (MonoTypeEnum type);
+static const int fast_log2 [] = { 1, 0, 1, -1, 2, -1, -1, -1, 3 };
```
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:556]: This is not simply calculating log2. It seems that you've assigned `-1` to places where you think should be illegal element numbers. If that's the case, element number 0 and 1 should be `-1` as well.

**Review by fanyang-mono — APPROVED** [review]

**Pull request merged**